editor/editor_command_palette.cpp

```diff
@@ -0,0 +1,284 @@
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "editor/editor_command_palette.h"
+#include "core/os/keyboard.h"
+#include "editor/editor_node.h"
+#include "scene/gui/tree.h"
+
+EditorCommandPalette *EditorCommandPalette::singleton = nullptr;
+
+float EditorCommandPalette::_score_path(const String &p_search, const String &p_path) {
+	float score = 0.9f + .1f * (p_search.length() / (float)p_path.length());
+
+	// Positive bias for matches close to the beginning of the file name.
+	int pos = p_path.findn(p_search);
+	if (pos != -1) {
+		return score * (1.0f - 0.1f * (float(pos) / p_path.length()));
+	}
+
+	// Positive bias for matches close to the end of the path.
+	pos = p_path.rfindn(p_search);
+	if (pos != -1) {
+		return score * (0.8f - 0.1f * (float(p_path.length() - pos) / p_path.length()));
+	}
+
+	// Remaining results belong to the same class of results.
+	return score * 0.69f;
+}
+
+void EditorCommandPalette::_update_command_search(const String &search_text) {
+	commands.get_key_list(&command_keys);
+	ERR_FAIL_COND(command_keys.is_empty());
+
+	// const String search_text = command_search_box->get_text();
+	const bool empty_search = search_text.is_empty();
+	Map<String, TreeItem *> sections;
+	TreeItem *first_section = nullptr;
+
+	// Filter possible candidates.
+	Vector<CommandEntry> entries;
+	for (int i = 0; i < command_keys.size(); i++) {
+		CommandEntry r;
+		r.key_name = command_keys[i];
+		r.display_name = commands[r.key_name].name;
+		r.shortcut_text = commands[r.key_name].shortcut;
+		if (!empty_search && search_text.is_subsequence_ofi(r.display_name)) {
+			r.score = _score_path(search_text, r.display_name.to_lower());
+			entries.push_back(r);
+		}
+	}
+
+	command_keys.clear();
+
+	TreeItem *root = search_options->get_root();
+	root->clear_children();
+
+	if (entries.size() > 0) {
+		if (!empty_search) {
+			SortArray<CommandEntry, CommandEntryComparator> sorter;
+			sorter.sort(entries.ptrw(), entries.size());
+		}
+
+		const int entry_limit = MIN(entries.size(), 300);
+		for (int i = 0; i < entry_limit; i++) {
+			String section_name = entries[i].key_name.get_slice("/", 0);
+			TreeItem *section;
+
+			if (sections.has(section_name)) {
+				section = sections[section_name];
+			} else {
+				section = search_options->create_item(root);
+
+				if (first_section == nullptr) {
```

Comment:
```suggestion
				if (!first_section) {
```
Matter of style preference, but comparing with `nullptr` is unnecessary and longer.

Reply:
thanks, changed.